src/hardware/vga_draw.cpp

```diff
@@ -1905,505 +1974,616 @@ void VGA_SetupDrawing(uint32_t /*val*/)
+			// a scaler.
+			if (is_vga_scan_doubling()) {
+				video_mode.height = vert_end / 2;
+				if (force_single_scan) {
+					vga.draw.address_line_total /= 2;
+					render_height = video_mode.height;
+					render_pixel_aspect_ratio /= 2;
+				} else {
+					render_height = video_mode.height * 2;
+				}
+			} else {
+				video_mode.height = vert_end;
+				render_height     = video_mode.height;
+			}
+			if (vga.seq.clocking_mode.is_pixel_doubling) {
+				render_pixel_aspect_ratio *= 2;
 			}
 		}
-		break;
 
-	case M_LIN8:
-		if (vga.crtc.mode_control.div_memory_address_clock_by_2) {
-			width >>= 1;
-		} else if (svgaCard == SVGA_S3Trio && !(vga.s3.reg_3a & 0x10)) {
-			doublewidth = true;
-			width >>= 1;
+		if (IS_VGA_ARCH) {
+			bpp = 32;
+
+			VGA_DrawLine = draw_linear_line_from_dac_palette;
+		} else {
+			VGA_DrawLine = VGA_Draw_Linear_Line;
 		}
-		[[fallthrough]];
-	case M_LIN24:
-	case M_LIN32:
-		width <<= 3;
-		if (vga.crtc.mode_control.div_memory_address_clock_by_2) {
-			doublewidth = true;
-			if (vga.mode == M_LIN32 && CurMode->mode == 0x10f) {
-				pixel_aspect_ratio /= 2;
+		break;
+
+	case M_TANDY16:
+		// 160x200 and 320x200 16-colour modes on Tandy & PCjr
+		vga.draw.blocks = horiz_end * 2;
+
+		if (vga.tandy.mode.is_high_bandwidth) {
+			if (machine == MCH_TANDY &&
+			    vga.tandy.mode.is_tandy_640_dot_graphics) {
+				vga.draw.blocks  = horiz_end * 4;
+				video_mode.width = horiz_end * 8;
+				render_width     = video_mode.width;
+			} else {
+				video_mode.width = horiz_end * 4;
+				render_width     = video_mode.width;
 			}
+			VGA_DrawLine = VGA_Draw_4BPP_Line;
+		} else {
+			video_mode.width = horiz_end * 4;
+			render_width     = video_mode.width * 2;
+
+			VGA_DrawLine = VGA_Draw_4BPP_Line_Double;
 		}
-		// Use HW mouse cursor drawer if enabled
-		bpp = VGA_ActivateHardwareCursor();
+
+		video_mode.height = vert_end;
+		render_height     = video_mode.height;
+
+		render_pixel_aspect_ratio = calc_pixel_aspect_from_dimensions(
+		        render_width, render_height, double_width, double_height);
 		break;
 
-	case M_LIN15:
-	case M_LIN16:
-		// 15/16 bpp modes double the horizontal values
-		width <<= 2;
-		if (vga.crtc.mode_control.div_memory_address_clock_by_2) {
-			doublewidth = true;
+	case M_TANDY4:
+		// 320x200 4-colour CGA mode on CGA, Tandy & PCjr
+		// 640x200 4-colour mode on Tandy & PCjr
+		vga.draw.blocks = horiz_end * 2;
+
+		video_mode.width  = horiz_end * 8;
+		video_mode.height = vert_end;
+
+		/*
+		        TODO This bit is not set for the 640x200
+		        Tandy mode; this cannot be correct. This would
+		        result in width-doubling the 640x200 mode. A
+		        simple 'width < 640' check will suffice instead
+		        until someone investigates this...
+
+		        if (machine == MCH_TANDY) {
+		        	double_width = (vga.tandy.mode.is_tand_640_dot_graphics) == 0;
+				} else { double_width =
+					(vga.tandy.mode.is_high_bandwidth) == 0;
+		        }
+		*/
```

Comment:
Comment for self: outdated comment, in the end I got rid of the `width < 640` check; so this should be removed.

Reply:
Actually, it has to stay. I'm doing `width < 640` later when I add the pixel-doubling stuff, so the comment is still relevant.